From 3776a2093052f5cc2bf826735a3e3d5b3e4edbf7 Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Thu, 23 Oct 2014 09:55:00 -0400
Subject: [fs] ext4: update PF_MEMALLOC handling in ext4_write_inode()

Message-id: <1414058125-4183-147-git-send-email-lczerner@redhat.com>
Patchwork-id: 98209
O-Subject: [RHEL 7.1 PATCH v2 146/171] ext4: update PF_MEMALLOC handling in ext4_write_inode()
Bugzilla: 1150139
RH-Acked-by: Eric Sandeen <sandeen@redhat.com>

From: Theodore Ts'o <tytso@mit.edu>

BZ 1150139
https://bugzilla.redhat.com/show_bug.cgi?id=1150139

BREW 8142710
https://brewweb.devel.redhat.com/taskinfo?taskID=8142710

Upstream commit 87f7e41636ff201148443551d06bc74497160aac

 The special handling of PF_MEMALLOC callers in ext4_write_inode()
 shouldn't be necessary as there shouldn't be any. Warn about it. Also
 update comment before the function as it seems somewhat outdated.

 (Changes modeled on an ext3 patch posted by Jan Kara to the linux-ext4
 mailing list on Februaryt 28, 2014, which apparently never went into
 the ext3 tree.)

 Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>
 Cc: Jan Kara <jack@suse.cz>
 Signed-off-by: Lukas Czerner <lczerner@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/ext4/inode.c b/fs/ext4/inode.c
index 30abd84..31d87bd 100644
--- a/fs/ext4/inode.c
+++ b/fs/ext4/inode.c
@@ -4436,21 +4436,20 @@ out_brelse:
  *
  * We are called from a few places:
  *
- * - Within generic_file_write() for O_SYNC files.
+ * - Within generic_file_aio_write() -> generic_write_sync() for O_SYNC files.
  *   Here, there will be no transaction running. We wait for any running
  *   transaction to commit.
  *
- * - Within sys_sync(), kupdate and such.
- *   We wait on commit, if tol to.
+ * - Within flush work (sys_sync(), kupdate and such).
+ *   We wait on commit, if told to.
  *
- * - Within prune_icache() (PF_MEMALLOC == true)
- *   Here we simply return.  We can't afford to block kswapd on the
- *   journal commit.
+ * - Within iput_final() -> write_inode_now()
+ *   We wait on commit, if told to.
  *
  * In all cases it is actually safe for us to return without doing anything,
  * because the inode has been copied into a raw inode buffer in
- * ext4_mark_inode_dirty().  This is a correctness thing for O_SYNC and for
- * knfsd.
+ * ext4_mark_inode_dirty().  This is a correctness thing for WB_SYNC_ALL
+ * writeback.
  *
  * Note that we are absolutely dependent upon all inode dirtiers doing the
  * right thing: they *must* call mark_inode_dirty() after dirtying info in
@@ -4462,15 +4461,15 @@ out_brelse:
  * stuff();
  * inode->i_size = expr;
  *
- * is in error because a kswapd-driven write_inode() could occur while
- * `stuff()' is running, and the new i_size will be lost.  Plus the inode
- * will no longer be on the superblock's dirty inode list.
+ * is in error because write_inode() could occur while `stuff()' is running,
+ * and the new i_size will be lost.  Plus the inode will no longer be on the
+ * superblock's dirty inode list.
  */
 int ext4_write_inode(struct inode *inode, struct writeback_control *wbc)
 {
  int err;
 
- if (current->flags & PF_MEMALLOC)
+ if (WARN_ON_ONCE(current->flags & PF_MEMALLOC))
   return 0;
 
  if (EXT4_SB(inode->i_sb)->s_journal) {
-- 
1.7.1